### ISSUE TITLE
Deadlock in CDomoticzHardwareBase::Stop()

### ISSUE THREAD

**Issue by emontnemery** [domoticz/domoticz#2671]
Taking the readQueueMutex in CDomoticzHardwareBase::Stop causes deadlocks when shutting down the HW class if there is some data arriving after taking the mutex and before the socket / serial port has been shut down.
```
bool CDomoticzHardwareBase::Stop()
{
	std::lock_guard<std::mutex> l(readQueueMutex); // Why do we need the mutex??
	return StopHardware();
}
```

Most HW classes have some pattern like this:
```
std::lock_guard<std::mutex> l(readQueueMutex); // Deadlock if the HW is being shut down!!
ParseData((const char*)pData,length);
```

To reproduce more easily, add a sleep() before the call to StopHardware:
```
bool CDomoticzHardwareBase::Stop()
{
	std::lock_guard<std::mutex> l(readQueueMutex); // Why do we need the mutex??
	sleep(10); // Make sure some data is received during the sleep to trigger deadlock
	return StopHardware();
}
```

**Comment by gizmocuz** [comment]
Should be fixed now. Only a few hardware indeed.

**Comment by emontnemery** [comment]
@gizmocuz Are you sure this is fixed? Please note #2668 did not fix this issue.

**Comment by gizmocuz** [comment]
@emontnemery , sorry , thought the above fix, fixed it

**Comment by gizmocuz** [comment]
So, when you need to close a hardware class, this needs to be signaled to the thread, once the thread knows it needs to shutdown, it needs (obviously), close serial/tcp readers, then exit nicely
This needs to be done in the thread

**Comment by emontnemery** [comment]
@gizmocuz Any idea what the purpose of the readQueueMutex in CDomoticzHardwareBase is or was?
To me it seems as if it is no longer needed and I want to remove it.

Edit: After some reviewing, it seems the purpose of the readQueueMutex is to serialize access to:
CDomoticzHardwareBase::m_rxbuffer and CDomoticzHardwareBase::m_rxbufferpos
However, m_rxbuffer is only used by two functions:
CDomoticzHardwareBase::onRFXMessage
CRFXBase::onInternalMessage

Hence, I would like to refactor such that:
readQueueMutex, m_rxbuffer, m_rxbufferpos, onRFXMessage are removed from CDomoticzHardwareBase
readQueueMutex, m_rxbuffer, m_rxbufferpos, onRFXMessage are added to CRFXBase
DomoticzTCP is modified such that it also extends CRFXBase
Thoughts?

**Comment by gizmocuz** [comment]
It had indeed something to do with the RFXCom implementation and was added quite some time ago.
That sounds like a good solution!

**Comment by emontnemery** [comment]
Fixed by #2680 + #2681
